src/System.Net.Sockets/tests/FunctionalTests/NetworkStreamTest.cs

```diff
@@ -11,6 +11,452 @@ namespace System.Net.Sockets.Tests
+        {
+            Assert.Throws<ArgumentNullException>("socket", () => new NetworkStream(null));
+            Assert.Throws<ArgumentNullException>("socket", () => new NetworkStream(null, false));
+            Assert.Throws<ArgumentNullException>("socket", () => new NetworkStream(null, true));
+            Assert.Throws<ArgumentNullException>("socket", () => new NetworkStream(null, FileAccess.ReadWrite));
+            Assert.Throws<ArgumentNullException>("socket", () => new NetworkStream(null, FileAccess.ReadWrite, false));
+        }
+
+        [Fact]
+        public void Ctor_NotConnected_ThrowsIOException()
+        {
+            Assert.Throws<IOException>(() => new NetworkStream(new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp)));
+        }
+
+        [Fact]
+        public async Task Ctor_NotStream_ThrowsIOException()
+        {
+            using (Socket listener = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp))
+            using (Socket client = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp))
+            {
+                listener.Bind(new IPEndPoint(IPAddress.Loopback, 0));
+                await client.ConnectAsync(new IPEndPoint(IPAddress.Loopback, ((IPEndPoint)listener.LocalEndPoint).Port));
+                Assert.Throws<IOException>(() => new NetworkStream(client));
+            }
+        }
+
+        [Fact]
+        public async Task Ctor_NonBlockingSocket_ThrowsIOException()
+        {
+            using (Socket listener = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
+            using (Socket client = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
+            {
+                listener.Bind(new IPEndPoint(IPAddress.Loopback, 0));
+                listener.Listen(1);
+
+                Task<Socket> acceptTask = listener.AcceptAsync();
+                await Task.WhenAll(acceptTask, client.ConnectAsync(new IPEndPoint(IPAddress.Loopback, ((IPEndPoint)listener.LocalEndPoint).Port)));
+                using (Socket server = await acceptTask)
+                {
+                    server.Blocking = false;
+                    Assert.Throws<IOException>(() => new NetworkStream(server));
+                }
+            }
+        }
+
+        [Fact]
+        public async Task Ctor_Socket_CanReadAndWrite_DoesntOwn()
+        {
+            using (Socket listener = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
+            using (Socket client = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
+            {
+                listener.Bind(new IPEndPoint(IPAddress.Loopback, 0));
+                listener.Listen(1);
+
+                Task<Socket> acceptTask = listener.AcceptAsync();
+                await Task.WhenAll(acceptTask, client.ConnectAsync(new IPEndPoint(IPAddress.Loopback, ((IPEndPoint)listener.LocalEndPoint).Port)));
```

Comment:
> then if the connect fails, the test fails.

And what if the accept fails?  Then there's still a hang.

If you need to be able to handle either failing or both succeeding, you need to do something like (the double await is on purpose):
```C#
await await Task.WhenAny(first, second);
await Task.WhenAll(first, second);
```
which is a more abbreviated version of:
https://github.com/dotnet/corefx/blob/bffef76f6af208e2042a2f27bc081ee908bb390b/src/System.Net.Http/tests/FunctionalTests/TestHelper.cs#L84

IIRC, we have a bunch of these WhenAlls in the tests.  If we're concerned about them, we should systematically replace them all with such a helper.